packages/@aws-cdk/aws-cloudwatch/lib/composite-alarm.ts

```diff
@@ -0,0 +1,241 @@
+import { Construct, Lazy, Resource, Stack } from '@aws-cdk/core';
+import { AlarmBase, AlarmState, IAlarm, IAlarmRule } from './alarm';
+import { CfnCompositeAlarm } from './cloudwatch.generated';
+
+/**
+ * Enumeration of supported Composite Alarms operators.
+ */
+enum Operator {
+
+  AND = 'AND',
+  OR = 'OR',
+  NOT = 'NOT',
+
+}
+
+/**
+ * Class with static functions to build AlarmRule for Composite Alarms.
+ */
+export class AlarmRule {
+
+  /**
+   * function to join all provided AlarmRules with AND operator.
+   *
+   * @param operands IAlarmRules to be joined with AND operator.
+   */
+  public static allOf(...operands: IAlarmRule[]): IAlarmRule {
+    return this.concat(Operator.AND, ...operands);
+  }
+
+  /**
+   * function to join all provided AlarmRules with OR operator.
+   *
+   * @param operands IAlarmRules to be joined with OR operator.
+   */
+  public static anyOf(...operands: IAlarmRule[]): IAlarmRule {
+    return this.concat(Operator.OR, ...operands);
+  }
+
+  /**
+   * function to wrap provided AlarmRule in NOT operator.
+   *
+   * @param operand IAlarmRule to be wrapped in NOT operator.
+   */
+  public static not(operand: IAlarmRule): IAlarmRule {
+    // tslint:disable-next-line:new-parens
+    return new class implements IAlarmRule {
+      public toAlarmRule(): string {
+        return `(NOT ${operand.toAlarmRule()})`;
+      }
+    };
+  }
+
+  /**
+   * function to build TRUE/FALSE intent for Rule Expression.
+   *
+   * @param value boolean value to be used in rule expression.
+   */
+  public static fromBoolean(value: boolean): IAlarmRule {
+    // tslint:disable-next-line:new-parens
+    return new class implements IAlarmRule {
+      public toAlarmRule(): string {
+        return `${String(value).toUpperCase()}`;
+      }
+    };
+  }
+
+  /**
+   * function to build Rule Expression for given IAlarm and AlarmState.
+   *
+   * @param alarm IAlarm to be used in Rule Expression.
+   * @param alarmState AlarmState to be used in Rule Expression.
+   */
+  public static fromAlarm(alarm: IAlarm, alarmState: AlarmState): IAlarmRule {
+    // tslint:disable-next-line:new-parens
+    return new class implements IAlarmRule {
+      public toAlarmRule(): string {
+        return `${alarmState}(${alarm.alarmArn})`;
+      }
+    };
+  }
+
+  /**
+   * function to build Rule Expression for given Alarm Rule string.
+   *
+   * @param alarmRule string to be used in Rule Expression.
+   */
+  public static fromString(alarmRule: string): IAlarmRule {
+    // tslint:disable-next-line:new-parens
+    return new class implements IAlarmRule {
+      public toAlarmRule(): string {
+        return alarmRule;
+      }
+    };
+  }
+
+  private static concat(operator: Operator, ...operands: IAlarmRule[]): IAlarmRule {
+    // tslint:disable-next-line:new-parens
+    return new class implements IAlarmRule {
+      public toAlarmRule(): string {
+        return operands
+          .map(operand => `(${operand.toAlarmRule()})`)
```

Comment:
This is parenthesizing unnecessarily. Instead of:

`(a) OR (b) OR (c)`

It should produce:

`(a OR b OR c)`

You probably did this because of:

```
concat(OR,
  a,
  concat(AND, b, c),
  d
)

# Right now leads to:
(a) OR ((b) AND (c)) OR (d)

# But equally correct with fewer parens is
(a OR (b AND C) OR d)
```

Am I missing an edge case?

Reply:
you're right, there was no easy means of solving for nesting without writing more code. the unnecessary parentheses solve for it quite conveniently at the expense of unnecessary tokens. i can go into implementing a duplicate parentheses remover, but that might not be trivial.

Reply:
Wouldn't just putting parens around the outside of the n-ary operator like I proposed solve it nicely?

Reply:
my apprehension with that is, i don't know if the operands are single tokens or expressions, they could be either. the same goes for the not operator, it could be a single token or an expression.

if the n-ary expressions are created only using the AlarmRule methods then i think we can be confident to a certain extent that they'll be accurately parenthesized.